README
Compression v1.4 - CSS Minifier and Caching
*******************************************

Requirements
============

	* PHP 5.2.x

Features
========

    * Can define variables within your stylesheet instead of typing the same value repeatedly
    * Can call PHP functions within the CSS to output dynamic stylesheets
    * Edit what delimiters are used for your variables
    * Stylesheets are compressed and stripped of whitespace
    * Logs a ratio of how much space was saved during compression
    * Enable or disable caching
    * Cached files and folders are created relative to the parent folder

Documentation
=============

	* http://www.milesj.me/resources/script/compression



docs/changelog.md
# Changelog #

*These logs may be outdated or incomplete.*

## 3.0.0 ##

* Updated to PHP 5.3
* Fixed Composer issues

## 2.1 ##

* Added Composer support
* Replaced errors with exceptions
* Refactored to use strict equality

## 2.0 ##

* Added support for stylesheets within subfolders
* Added support for calling class methods within CSS: @ClassName.methodName()
* Changed CSS variables and functions to require an @ before the name
* Converted private members to protected
* Improved the caching mechanism

## 1.4 ##

* Added support for inline CSS functions, written as PHP functions
* Added the parseFunctions() method to handle the new CSS function mechanism
* Added a disableCache() method to be called to disable browser caching
* Added a setCachePath() method so you can define your own cached files location
* Rewrote __construct() with faster logic
* 

## 1.3 ##

* First initial release of Compression


tests/index.php
<?php
/**
 * @copyright	Copyright 2006-2012, Miles Johnson - http://milesj.me
 * @license		http://opensource.org/licenses/mit-license.php - Licensed under the MIT License
 * @link		http://milesj.me/code/php/compression
 */

// Define our dynamic function
// Can be used in the CSS file as @colWidth()
function colWidth($size, $base = 100) {
	return ($size * $base) .'px';
}

// Get the path from the query: index.php?load=style.css,sub/style.css
// Separate multiple stylesheets with a comma
$stylesheet = isset($_GET['load']) ? $_GET['load'] : 'style.css';

// Include class and instantiate
include_once '../Compression.php';

// Accepts a comma separated string, or an array
$css = new \mjohnson\compression\Compression($stylesheet);

// Set the location of the CSS files; trailing slash optional
$css->setPath(dirname(__FILE__) . '/css/');

// Turn caching off for testing purposes
$css->setCaching(false);

// Bind the variables and parse
// Variables can be used as @variableName
$css->bind(array(
	'font_family' => '"Verdana", "Arial", sans-serif',
	'blue' => '#0000FF',
	'img' => '/images'
));

// Output the compressed version
echo $css->parse();


tests/example.php
<?php
/**
 * Example file to load CSS files through Compression
 * css.php?load=style.css
 */

// Get the path from the query
$stylesheet = isset($_GET['load']) ? $_GET['load'] : 'style.css';

// Define our custom function!
function colWidth() {
	$args = func_get_args(); 
	$width = $args[0] * 100;
	return $width .'px';
}

// Require class
require_once('compression.php');

// Initialize
$css = new Compression($stylesheet);

// Set the location of the css files; trailing slash optional
$css->setPath(dirname(__FILE__) . DIRECTORY_SEPARATOR .'css', 'cache');

// Turn cache off for testing purposes
$css->setCaching(false);

// Bind the variables and parse
$css->bind(array(
	'img' 	=> '/images/',
	'font' 	=> '"Verdana", "Arial", sans-serif',
	'blue'	=> '#0000FF'
));

// Output the compressed version
$css->parse();


Compression.php
<?php
/**
 * @copyright	Copyright 2006-2012, Miles Johnson - http://milesj.me
 * @license		http://opensource.org/licenses/mit-license.php - Licensed under the MIT License
 * @link		http://milesj.me/code/php/compression
 */

namespace mjohnson\compression;

use \Exception;

/**
 * A basic class that loads CSS files, compresses them, and saves the resulting output into a cached file.
 * Also supports dynamic variables and functions within the CSS file.
 *
 * @version	3.0.1
 * @package	mjohnson.compression
 */
class Compression {

	/**
	 * Is caching enabled?
	 *
	 * @access protected
	 * @var boolean
	 */
	protected $_cache = true;

	/**
	 * Path to the cached files, relative to the given CSS path.
	 *
	 * @access protected
	 * @var string
	 */
	protected $_cachePath;

	/**
	 * The paths of the CSS files.
	 *
	 * @access protected
	 * @var array
	 */
	protected $_css = array();

	/**
	 * The path to the directory holding the CSS files.
	 *
	 * @access protected
	 * @var string
	 */
	protected $_cssPath;

	/**
	 * Should the stylesheet be parsed?
	 *
	 * @access protected
	 * @var boolean
	 */
	protected $_parse = true;

	/**
	 * Array of CSS variable references.
	 *
	 * @access protected
	 * @var array
	 */
	protected $_variables = array();

	/**
	 * Loads the CSS file into the class.
	 *
	 * @access public
	 * @param string|array $stylesheets
	 */
	public function __construct($stylesheets = array()) {
		if ($stylesheets) {
			if (is_string($stylesheets)) {
				$stylesheets = explode(',', trim($stylesheets));
			}

			foreach ($stylesheets as $sheet) {
				if (strtolower(substr(strrchr($sheet, '.'), 1)) !== 'css') {
					$sheet = trim($sheet, '.') .'.css';
				}

				$this->_css[] = trim($sheet, '/');
				$this->_variables = array();
			}

			$this->setPath();
		} else {
			$this->_parse = false;
		}
	}

	/**
	 * Binds variables to the CSS stylesheets.
	 *
	 * @access public
	 * @param string|array $variable
	 * @param string $value
	 * @return Compression
	 * @chainable
	 */
	public function bind($variable, $value = null) {
		if (!$this->_parse) {
			return $this;
		}

		if (is_array($variable)) {
			foreach ($variable as $var => $value) {
				$this->bind($var, $value);
			}
		} else {
			$variable = preg_replace('/[^-_a-zA-Z0-9]/i', '', $variable);

			if (substr($variable, 0, 1) !== '@') {
				$variable = '@' . $variable;
			}

			$this->_variables[$variable] = trim(strip_tags($value));
		}

		return $this;
	}

	/**
	 * Forces the users browser not to cache the results of the current request.
	 *
	 * @access public
	 * @return void
	 */
	public function disableCache() {
		header("Expires: Mon, 26 Jul 1997 05:00:00 GMT");
		header("Last-Modified: " . gmdate("D, d M Y H:i:s") . " GMT");
		header("Cache-Control: no-store, no-cache, must-revalidate");
		header("Cache-Control: post-check=0, pre-check=0", false);
		header("Pragma: no-cache");
	}

	/**
	 * Parses the stylesheet; does important logic for caching and compressing.
	 *
	 * @access public
	 * @return string
	 */
	public function parse() {
		$response = "";
		$cssModified = time();

		if (!$this->_parse || !$this->_css) {
			return $response;
		}

		foreach ($this->_css as $css) {
			$baseCss = $this->_cssPath . $css;
			$cachedCss = $this->_cachePath . $css;
			$cache = true;
			$output = "";

			// Use cache or regenerate
			if (is_file($cachedCss)) {
				$cssModified = filemtime($baseCss);

				if ($cssModified > filemtime($cachedCss)) {
					$output = $this->_compress($baseCss, $css);
				} else {
					$output = file_get_contents($cachedCss);
					$cache = false;
				}

			// Use base CSS
			} else if (is_file($baseCss)) {
				$output = $this->_compress($baseCss, $css);
				$cssModified = time();

			// No CSS
			} else {
				continue;
			}

			if ($this->_cache && $cache){
				$this->_cache($css, $output);
			}

			$response .= $output . "\n\n";
		}

		header("Date: " . date("D, d M Y H:i:s T", $cssModified) . " GMT");
		header("Content-Type: text/css");
		header("Expires: " . gmdate("D, d M Y H:i:s T", time() + 86400) . " GMT");
		header("Cache-Control: max-age=86400, must-revalidate"); // HTTP/1.1
		header("Pragma: cache"); // HTTP/1.0

		return $response;
	}

	/**
	 * Enable or disable caching.
	 *
	 * @access public
	 * @param boolean $enable
	 * @return Compression
	 * @chainable
	 */
	public function setCaching($enable = true) {
		$this->_cache = (boolean) $enable;

		return $this;
	}

	/**
	 * Set the path to the location of the CSS files (and cached files).
	 *
	 * @access public
	 * @param string $path
	 * @param string $cachePath
	 * @return Compression
	 * @chainable
	 */
	public function setPath($path = null, $cachePath = '_cache') {
		if (!$path) {
			$path = dirname(__FILE__);
		}

		$path = trim(str_replace('\\', '/', $path), '/');

		if (substr($path, -1) !== '/') {
			$path .= '/';
		}

		$this->_cssPath = $path;
		$this->_cachePath = $path . $cachePath . '/';

		return $this;
	}

	/**
	 * Creates a cached file of the CSS.
	 *
	 * @access protected
	 * @param string $name
	 * @param string $content
	 * @return void
	 */
	protected function _cache($name, $content) {
		$path = $this->_cachePath . $name;
		$dir = dirname($path);

		if (!is_dir($dir)) {
			mkdir($dir, 0777);

		} else if (!is_writeable($dir)) {
			chmod($dir, 0777);
		}

		file_put_contents($path, $content);
	}

	/**
	 * Compress the CSS and bind the variables.
	 *
	 * @access protected
	 * @param string $path
	 * @param string $name
	 * @return string
	 */
	protected function _compress($path, $name) {
		$stylesheet = file_get_contents($path);

		// Parse the functions
		$stylesheet = preg_replace_callback('/(?:@([_\.a-zA-Z0-9]+)\((.*?)\))/i', array($this, '_functionize'), $stylesheet);

		// Parse the variables
		if ($this->_variables) {
			$stylesheet = str_replace(array_keys($this->_variables), $this->_variables, $stylesheet);
		}

		// Remove all whitespace
		$output = preg_replace('!/\*[^*]*\*+([^/][^*]*\*+)*/!', '', $stylesheet);
		$output = str_replace(array("\r\n", "\r", "\n", "\t", '/\s\s+/', '  ', '   '), '', $output);
		$output = str_replace(array(' {', '{ '), '{', $output);
		$output = str_replace(array(' }', '} '), '}', $output);
		$output = str_replace(': ', ':', $output);

		$ratio  = 100 - (round(strlen($output) / strlen($stylesheet), 3) * 100);
		$output = "/* $name ($ratio%) */\n". $output;

		return $output;
	}

	/**
	 * Parses the document and execute custom inline functions.
	 *
	 * @access protected
	 * @param string $matches
	 * @return string
	 * @throws \Exception
	 */
	protected function _functionize($matches) {
		$function = str_replace('@', '', trim($matches[1]));
		$args = !empty($matches[2]) ? array_map('trim', explode(',', $matches[2])) : $matches[2];

		if (strpos($function, '.') !== false) {
			list($class, $method) = explode('.', $function);

			if (method_exists($class, $method)) {
				return call_user_func_array(array($class, $method), $args);
			}

		} else if (function_exists($function)) {
			return call_user_func_array($function, $args);
		}

		throw new Exception(sprintf('Function %s does not exist.', $function));
	}

}


compression.php
<?php
/**
 * Compression - CSS Builder, Compressor and Cacher
 *
 * Allows the use of defined variables within the CSS file; also compresses the stylesheet and caches it.
 *
 * @author       Miles Johnson - www.milesj.me
 * @copyright    Copyright 2006-2009, Miles Johnson, Inc.
 * @license      http://www.opensource.org/licenses/mit-license.php - Licensed under The MIT License
 * @link         www.milesj.me/resources/script/compression
 */

class Compression {

    /**
     * Current version: www.milesj.me/resources/logs/compression
     *
     * @access public
     * @var int
     */
    public $version = '1.5';

    /**
     * Is cacheing enabled?
     *
     * @access private
     * @var boolean
     */
    private $__cache = true;

    /**
     * Path to the cached files, relative to the given css.
     *
     * @access private
     * @var string
     */
    private $__cachePath;

    /**
     * The path to the css file.
     *
     * @access private
     * @var array
     */
    private $__css = array();

    /**
     * The path to the directory holding the css files.
     *
     * @access private
     * @var string
     */
    private $__cssPath;

    /**
     * Should the stylesheet be parsed?
     *
     * @access private
     * @var boolean
     */
    private $__parse = true;

    /**
     * Array of css variable references.
     *
     * @access private
     * @var array
     */
    private $__variables;

    /**
     * The prefix delimiter of your variable.
     *
     * @access private
     * @var string
     */
    private $__varPre = '[';

    /**
     * The suffix delimiter of your variable.
     *
     * @access private
     * @var string
     */
    private $__varSuf = ']';

    /**
     * Loads the css file into the class.
     *
     * @access public
     * @param string $stylesheets
     * @return void
     */
    public function __construct($stylesheets = array()) {
        if (!is_array($stylesheets)) {
            $stylesheets = explode(',', str_replace("\s", '', $stylesheets));
        }

        if (!empty($stylesheets)) {
            foreach ($stylesheets as $sheet) {
                if (mb_strtolower(substr(strrchr($sheet, '.'), 1)) != 'css') {
                    $sheet = trim($sheet, '.') .'.css';
                }

                $this->__css[] = basename($sheet);
                $this->__variables = array();
            }
        } else {
            trigger_error('Compression::__construct(): No stylesheets have been defined', E_USER_WARNING);
            $this->__parse = false;
        }
    }

    /**
     * Binds variables to the CSS stylesheet.
     *
     * @access public
     * @param string $variable
     * @param string $value
     * @return object
     */
    public function bind($variable, $value = null) {
        if ($this->__parse === false) {
            return;
        }

        if (is_array($variable)) {
            foreach ($variable as $var => $value) {
                $this->bind($var, $value);
            }
        } else {
            if (!empty($variable) && !empty($value)) {
                $variable = preg_replace('/[^-_a-zA-Z0-9]/i', '', $variable);

                if (substr($variable, 0, 1) != $this->__varPre) {
                    $variable = $this->__varPre . $variable;
                }

                if (substr($variable, -1) != $this->__varSuf) {
                    $variable = $variable . $this->__varSuf;
                }

                $this->__variables[$variable] = trim(htmlentities(strip_tags($value), ENT_NOQUOTES, 'UTF-8'));
            }
        }

        return $this;
    }

    /**
     * Forces the user's browser not to cache the results of the current request.
     *
     * @access public
     * @return void
     */
    public function disableCache() {
        header("Expires: Mon, 26 Jul 1997 05:00:00 GMT");
        header("Last-Modified: ". gmdate("D, d M Y H:i:s") ." GMT");
        header("Cache-Control: no-store, no-cache, must-revalidate");
        header("Cache-Control: post-check=0, pre-check=0", false);
        header("Pragma: no-cache");
    }

    /**
     * Parses the stylesheet; does required logic for cacheing and compressing.
     *
     * @access public
     * @param boolean $return
     * @return string
     */
    public function parse($return = false) {
        if ($this->__parse === false) {
            return;
        }

        $mainOutput = "";

        foreach ($this->__css as $css) {
            $baseCss = $this->__cssPath . $css;
            $cachedCss = $this->__cachePath . $css;
            $cache = true;
            $output = "";

            if (file_exists($cachedCss)) {
                $cssModified = filemtime($baseCss);
                $cacheModified = filemtime($cachedCss);

                if ($cssModified > $cacheModified) {
                    $output = $this->__compress($baseCss);
                } else {
                    $output = file_get_contents($cachedCss);
                    $cache = false;
                }
            } else if (file_exists($baseCss)) {
                $output = $this->__compress($baseCss);
                $cssModified = time();
            } else {
                trigger_error('Compression::parse(): Stylesheet "'. basename($baseCss) .'" could not be found', E_USER_WARNING);
            }

            if ($this->__cache && $cache){
                $this->__cache($css, $output);
            }

            $mainOutput .= $output;
            unset($cache, $output, $baseCss, $cachedCss);
        }

        if ($return) {
            return $mainOutput;
        } else {
            header("Date: ". date("D, j M Y G:i:s ", $cssModified) ." GMT");
            header("Content-Type: text/css");
            header("Expires: ". gmdate("D, j M Y H:i:s", time() + 86400) ." GMT");
            header("Cache-Control: max-age=86400, must-revalidate"); // HTTP/1.1
            header("Pragma: cache"); // HTTP/1.0
            echo $mainOutput;
        }
    }

    /**
     * Is cacheing enabled or disabled?
     *
     * @access public
     * @param boolean $enable
     * @return object
     */
    public function setCaching($enable = true) {
        if (is_bool($enable)) {
            $this->__cache = $enable;
        }

        return $this;
    }

    /**
     * Set the delimiters to use for the inline variables.
     *
     * @access public
     * @param string $prefix
     * @param string $suffix
     * @return object
     */
    public function setDelimiters($prefix = '[', $suffix = ']') {
        if (empty($prefix) && empty($suffix)) {
            return false;
        }

        $prefix = preg_replace('/[^-_=+;:<>{}\[\]|]/i', '', $prefix);
        if ($prefix != '') {
            $this->__varPre = $prefix;
        }

        $suffix = preg_replace('/[^-_=+;:<>{}\[\]|]/i', '', $suffix);
        if ($suffix != '') {
            $this->__varSuf = $suffix;
        }

        return $this;
    }

    /**
     * Set the path to the location of the css files (and cached files).
     *
     * @access public
     * @param string $path
     * @param string $cacheDir
     * @return object
     */
    public function setPath($path, $cacheDir = 'cache') {
        $root = $_SERVER['DOCUMENT_ROOT'];

        if (empty($path)) {
            $path = dirname(__FILE__);
        }

        $path = str_replace('\\', '/', $path);

        if (substr($path, 0, strlen($root)) != $root) {
            $path = $root . $path;
        }

        if (substr($path, -1) != DIRECTORY_SEPARATOR) {
            $path .= DIRECTORY_SEPARATOR;
        }

        $this->__cssPath = $path;
        $this->__cachePath = $path . $cacheDir . DIRECTORY_SEPARATOR;

        return $this;
    }

    /**
     * Creates a cached file of the CSS.
     *
     * @access private
     * @param string $css
     * @param string $input
     * @return string
     */
    private function __cache($css, $input) {
        if (!is_dir($this->__cachePath)) {
            mkdir($this->__cachePath, 0777);
        }

        $handle = fopen($this->__cachePath . basename($css), 'w');
        fwrite($handle, $input);
        fclose($handle);

        return true;
    }

    /**
     * Compress the CSS and bind the variables.
     *
     * @access private
     * @param string $css
     * @return string
     */
    private function __compress($css) {
        $stylesheet = file_get_contents($css);

        // Parse the variables
        if (!empty($this->__variables)) {
            $stylesheet = str_replace(array_keys($this->__variables), array_values($this->__variables), $stylesheet);
        }

        // Parse the functions
        $stylesheet = preg_replace_callback('/(?:([_a-zA-Z0-9]+)\((.*?)\))/i', array($this, '__functionize'), $stylesheet);

        // Remove all whitespace
        $output = preg_replace('!/\*[^*]*\*+([^/][^*]*\*+)*/!', '', $stylesheet);
        $output = str_replace(array("\r\n", "\r", "\n", "\t", '/\s\s+/', '  ', '   '), '', $output);
        $output = str_replace(array(' {', '{ '), '{', $output);
        $output = str_replace(array(' }', '} '), '}', $output);
        $output = str_replace(': ', ':', $output);

        $ratio  = 100 - (round(mb_strlen($output) / mb_strlen($stylesheet), 3) * 100);
        $output = "/* file: ". basename($css) .", ratio: $ratio% */ ". $output;

        return $output;
    }

    /**
     * Parses the document and runs custom inline functions.
     *
     * @access private
     * @param string $matches
     * @return string
     */
    private function __functionize($matches) {
        $function = $matches[1];
        $args = !empty($matches[2]) ? array_map('trim', explode(',', $matches[2])) : $matches[2];

        // Dont mess with existent css functions
        if (in_array($function, array('url', 'attr', 'rect', 'rgb', 'alpha', 'lang'))) {
            return $matches[0];
        } else {
            if (function_exists($function)) {
                return call_user_func_array($function, $args);
            } else {
                trigger_error('Compression::__functionize(): Custom function "'. $function .'" does not exist', E_USER_WARNING);
            }
        }

        return null;
    }

}
